core/modulesource.go

```diff
@@ -245,6 +249,150 @@ func (src *ModuleSource) AutomaticGitignore(ctx context.Context) (*bool, error)
 	return modCfg.Codegen.AutomaticGitignore, nil
 }
 
+// LoadContext loads a directory from the module context directory.
+//
+// If the module is local, it will load the directory from the local source
+// directly from the host.
+//
+// If the module is git, it will load the directory from the git repository
+// using its context directory.
+//
+// The path is relative to the module context directory, even if it's absolute.
+//
+// Improvement for later: use `.gitignore` to excludes paths from the context.
+func (src *ModuleSource) LoadContext(ctx context.Context, dag *dagql.Server, path string, ignore []string) (inst dagql.Instance[*Directory], err error) {
+	// We exclude .git by default because it's not useful and tends to invalidate the cache.
+	excludes := []string{"**/.git"}
+	includes := []string{}
+
+	for _, p := range ignore {
+		if strings.HasPrefix(p, "!") {
+			includes = append(includes, p[1:])
+		} else {
+			excludes = append(excludes, p)
+		}
+	}
+
+	// If path is not absolute, it's relative to the module root directory.
+	if !filepath.IsAbs(path) {
+		moduleRootPath, err := src.SourceRootSubpath()
+		if err != nil {
+			return inst, fmt.Errorf("failed to get source root subpath for relative context path: %w", err)
+		}
+
+		path = filepath.Join(moduleRootPath, path)
+	}
+
+	switch src.Kind {
+	case ModuleSourceKindLocal:
+		bk, err := src.Query.Buildkit(ctx)
+		if err != nil {
+			return inst, fmt.Errorf("failed to get buildkit api: %w", err)
+		}
+
+		slog.Debug("moduleSource.LoadContext: loading contextual directory from local source", "path", path, "kind", src.Kind)
+
+		ctxPath, err := src.resolveContextPath(ctx, bk)
+		if err != nil {
+			return inst, fmt.Errorf("failed to resolve context path: %w", err)
+		}
+
+		// Add the context path to the path to load
+		path = filepath.Join(ctxPath, path)
+
+		path, err = filepath.Rel(ctxPath, path)
+		if err != nil {
+			return inst, fmt.Errorf("path should be relative to the context path: %w", err)
+		}
+
+		_, desc, err := bk.LocalImport(ctx, src.Query.Platform().Spec(), path, excludes, includes)
+		if err != nil {
+			return inst, fmt.Errorf("failed to import local module src: %w", err)
+		}
+
+		inst, err = LoadBlob(ctx, dag, desc)
+		if err != nil {
+			return inst, fmt.Errorf("failed to load local module src: %w", err)
+		}
+
+		return inst, nil
+	case ModuleSourceKindGit:
+		slog.Debug("moduleSource.LoadContext: loading contextual directory from git", "path", path, "kind", src.Kind, "repo", src.AsGitSource.Value.HTMLURL())
+
+		// Use the Git context directory.
+		ctxDir := src.AsGitSource.Value.ContextDirectory.Self
+
+		dir, err := ctxDir.Directory(ctx, path)
+		if err != nil {
+			return inst, fmt.Errorf("failed to load contextual directory %q: %w", path, err)
+		}
+
+		loadedDir, err := dir.AsBlob(ctx, dag)
+		if err != nil {
+			return inst, fmt.Errorf("failed to get dir instance: %w", err)
+		}
```

Comment:
This uses one more blob conversion than neccessary - you still need the `Instance`, but you can get that by using `dag.Select`.

Reply:
There's an argument for doing blob conversion here in that it means contextual directories with equal contents will always be cached the same no matter their source. This can actually matter in cases like:
1. Local and git have same contents (i.e. local checkout of a git repo)
2. Getting the same directory from different git commits and/or forks of the same repo
3. Commonly reoccurring directories, "empty dir" being the most common I'd guess

But there's also overhead to blob-ification, so hard to say one is better than the other. If forced to choose I'd probably just leave this as is for now though.